### ISSUE TITLE
PNPM fails to detect invalid lockfile if `resolution.directory` is missing

### ISSUE THREAD

**Issue by gerbyzation** [pnpm/pnpm#6811]
### pnpm version:
v8.6.2

### Code to reproduce the issue:
https://github.com/gerbyzation/turbo-prune-pnpm-resolution-directory-missing-repro

### Expected behavior: 
PNPM detects the invalid lockfile and raise a descriptive error as to why the lockfile is invalid.

### Actual behavior:
Throws an error that doesn't make clear to the user what the problem is:
```
❯ pnpm i
Scope: all 3 workspace projects
Lockfile is up to date, resolution step is skipped
Packages: +8
++++++++
 ERR_INVALID_ARG_TYPE  The "path" argument must be of type string. Received undefined

pnpm [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:372:5)
    at validateString (node:internal/validators:120:11)
    at Object.join (node:path:1172:7)
    at directoryFetcher (/Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:102008:36)
    at fetcher (/Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:125428:22)
    at ctx.requestsQueue.add.priority.priority (/Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:125324:78)
    at run (/Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:124809:90)
    at PQueue._tryToStartAnother (/Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:124757:13)
    at /Users/gerbenneven/.fnm/node-versions/v16.15.1/installation/lib/node_modules/pnpm/dist/pnpm.cjs:124822:16
    at new Promise (<anonymous>)
Packages are copied from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/gerbenneven/Library/pnpm/store/v3
  Virtual store is at:             node_modules/.pnpm
Progress: resolved 1, reused 0, downloaded 0, added 0
```

### Additional information:
The core of the issue of course is that `turbo prune` is generating an invalid lockfile (I've filed an issue: https://github.com/vercel/turbo/issues/5529), which needs to be fixed on their part. 

In summary what turbo does wrong is it generates this: (note the missing `directory` field in `resolution`)
```yaml
  file:libraries/lib-b:
    resolution:
      type: directory
```
Instead of this:

```yaml
  file:libraries/lib-b:
    resolution: {directory: libraries/lib-b, type: directory}
```

 - `node -v` prints: v16.15.1 + v18
 - Windows, macOS, or Linux?: observed on mac + linux

**Comment by KSXGitHub** [comment]
@zkochan `zod` could be useful to prevent similar problems in the feature.

**Comment by zkochan** [comment]
We can try zod but I hope the additional processing won't add too much time to the installation time.

**Comment by aidenZou** [comment]
`pnpm add @types/big.js -D`

**Comment by zkochan** [comment]
This no longer reproduces on current pnpm, so closing.

Verified on pnpm 12.5.1: a lockfile whose resolution is missing a field fails with ERR_PNPM_BROKEN_LOCKFILE, naming the lockfile and the position of the bad entry.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
